### PR TITLE
Added failover.

### PR DESCRIPTION
**Pull request by CodeReaper**
When a FQDN have multiple DNS A records, these can serve as failover addresses. E.g. if one fails, try another.

This pull request contains changes to implement opt-in failover. You can opt-in a with a delegate callback which is checked for each address prior to failing over.

### CODE REVIEW

**Comment by chrisballinger** [comment]
Wouldn't it make more sense to use a property instead of a delegate? Just make sure to use thread safe get/set similar to `- (id)userData` and `- (void)setUserData:(id)arbitraryUserData`.

Also I think I will be merging the IPv6 branch https://github.com/robbiehanson/CocoaAsyncSocket/pull/430 soon, which will probably have merge conflicts with this.

Perhaps the failover mechanism from the IPv6 pull request could be re-used for this implementation, or they could be unified in some way.

**Comment by CodeReaper** [comment]
I did consider using a property, but I wanted to be able to implement a delegate for a host that might have 10 addresses, but only the first two would be attempted to fail over to. So I went with a delegate so people can build their own custom logic or just return YES.

Hmm... Maybe I should wait for the IPv6 branch to be merged and see how my pull request is doing. Do you have an idea about when that merge will happen?

**Comment by chrisballinger** [comment]
I was waiting for the bugs to get worked out, and it seems like at least one major one has been found so far and fixed. If no news means good news.. I'll probably merge it soon.

**Comment by CodeReaper** [comment]
I have tried patching the ipv6 branch with my pull request and I can see there are conflicts, but it seems fixable. Let me know you have merged the ipv6 branch and I will see look into getting failover back in merge-able state.

**Comment by CodeReaper** [comment]
@chrisballinger - I finally got back to trying to add failover once again. Hopefully you or someone else will look it over and tell me if I need to make some changes to it. :)

File: Source/GCD/GCDAsyncSocket.h [hunk 1]
```diff
@@ -128,12 +128,14 @@ typedef NS_ENUM(NSInteger, GCDAsyncSocketError) {
...
+ * When connecting to a FQDN which have multiple DNS A records this is the timeout
+ * for connecting to each given record.
+ * *NOTE* This is a different value than the one given to -(BOOL)connectTo...withTimeout(NSTimeInterval)timeout
+ * *NOTE* That records times timeout can be less than the value for the overall timeout.
  *
- * Defaults to 300ms.
+ * Defaults to 2s.
 **/
-@property (atomic, assign, readwrite) NSTimeInterval alternateAddressDelay;
+@property (atomic, assign, readwrite) int connectionAttemptTimeout;
```
Comment (chrisballinger) [inline Source/GCD/GCDAsyncSocket.h]: This variable can't be changed due to backwards compatibility

**Comment by chrisballinger** [comment]
Also why not do the same for multiple AAAA records?

The PR is pretty large (deletes large segments of code and re-adds other large segments) and not well documented so it's hard for me to quickly examine it for correctness.

**Comment by CodeReaper** [comment]
> Also why not do the same for multiple AAAA records?
Did I miss something is it not only the -connectToHost... methods that resolve addresses? The other all use an already resolved address or is a unix socket url?

> The PR is pretty large (deletes large segments of code and re-adds other large segments) and not well documented so it's hard for me to quickly examine it for correctness.
By not well documented, are you referring to the one line comments that are in other segments of code or something else?

**Pull request closed without merging**